### PR TITLE
Add exceptions reporting for source

### PR DESCRIPTION
**Pull request by CFarcy**

### CODE REVIEW

File: django_geosource/models.py [hunk 1]
```diff
@@ -129,26 +131,39 @@ def refresh_data(self):
...
                 except KeyError:
-                    raise Exception(
-                        "Can't find identifier field in one or more records"
+                    msg = "Can't find identifier field for this record"
+                    self.report["status"] = "Warning"
+                    self.report.setdefault("message", []).append(msg)
+                    self.report.setdefault("lines", {}).setdefault(f"{i}", []).append(
+                        msg
                     )
+                    self.save(update_fields=["report"])
```
Comment (jrmi) [inline django_geosource/models.py:147]: Could be very time consuming, just construct the report object, then save it.

File: django_geosource/models.py [hunk 2]
```diff
@@ -129,26 +131,39 @@ def refresh_data(self):
             layer = self.get_layer()
             begin_date = datetime.now()
             row_count = 0
```
Comment (jrmi) [inline django_geosource/models.py:132]: Create an empty report here and add item to it lateron.

File: django_geosource/models.py [hunk 3]
```diff
@@ -185,9 +200,15 @@ def update_fields(self):
                         try:
                             value = value.decode()
                         except (UnicodeDecodeError, AttributeError):
-                            logger.warning(
+                            msg = (
                                 f"{field_name} couldn't be decoded for source {self.pk}"
```
Comment (jrmi) [inline django_geosource/models.py:203]: Pk is not really human readable. Use name (beware of encoding, test with accentuated source name)

File: django_geosource/models.py [hunk 4]
```diff
@@ -297,7 +324,11 @@ def get_file_as_dict(self):
         try:
             return json.load(self.file)
         except json.JSONDecodeError:
-            logger.info("Source's GeoJSON file is not valid")
+            msg = "Source's GeoJSON file is not valid"
+            self.report["status"] = "Error"
+            self.report.setdefault("message", []).append(msg)
+            self.save()
+            logger.info(msg)
```
Comment (jrmi) [inline django_geosource/models.py:330]: You want to keep the logger info here ?

File: django_geosource/models.py [hunk 5]
```diff
@@ -317,9 +348,11 @@ def _get_records(self, limit=None):
                     }
                 )
             except (ValueError, GDALException):
-                raise ValueError(
-                    f"One of source's record has bad geometry: {record['geometry']}"
-                )
+                msg = f"The record has bad geometry: {record['geometry']}"
```
Comment (jrmi) [inline django_geosource/models.py:350]:
```suggestion
                msg = f"The record geometry seems invalid."
```

File: django_geosource/models.py [hunk 6]
```diff
@@ -406,10 +439,21 @@ def get_file_as_sheet(self):
                 encoding=self.settings["encoding"],
                 quotechar=quotechar,
             )
-        except pyexcel.exceptions.FileTypeNotSupported as err:
+        # Exception is raised if no parser found
+        except (pyexcel.exceptions.FileTypeNotSupported) as err:
             msg = "Source's CSV file is not valid"
```
Comment (jrmi) [inline django_geosource/models.py:443]:
```suggestion
            msg = "Provided CSV file seems invalid"
```

File: django_geosource/models.py [hunk 7]
```diff
@@ -445,23 +497,48 @@ def _get_records(self, limit=None):
...
+                records.append(
+                    {
+                        self.SOURCE_GEOM_ATTRIBUTE: GEOSGeometry(
+                            f"Point({x} {y})", srid=srid
+                        ),
+                        **cells,
+                    }
+                )
+            except (ValueError, GDALException):
+                msg = f"One of source's record has bad geometry: Point({x} {y}) srid={srid}"
```
Comment (jrmi) [inline django_geosource/models.py:509]: Can't we do it by line ?

Comment (jrmi) [inline django_geosource/models.py:509]:
```suggestion
                msg = f"One of source's record has invalid geometry: Point({x} {y}) srid={srid}"
```

File: django_geosource/models.py [hunk 8]
```diff
@@ -445,23 +497,48 @@ def _get_records(self, limit=None):
...
+                logger.info(msg)
```
Comment (jrmi) [inline django_geosource/models.py:510]: Keep this ?

File: django_geosource/models.py [hunk 9]
```diff
@@ -445,23 +497,48 @@ def _get_records(self, limit=None):
...
+                self.report["status"] = "Warning"
+                self.report.setdefault("message", []).append(msg)
+                self.save()
+                # raise ValueError(msg)
+                continue
+            row_count += 1
+        if not row_count:
+            self.report["status"] = "Error"
```
Comment (jrmi) [inline django_geosource/models.py:518]: Add a message: `No record could be imported, check your file`

File: django_geosource/models.py [hunk 10]
```diff
@@ -445,23 +497,48 @@ def _get_records(self, limit=None):
...
-            )
+            try:
+                field_index = (
+                    colnames.index(field)
+                    if self.settings.get("use_header")
+                    else int(field)
+                )
+            except ValueError as err:
+                msg = f"{field} is not a valid coordinate field"
+                logger.info(msg)
```
Comment (jrmi) [inline django_geosource/models.py:535]: Keep this ?

File: django_geosource/models.py [hunk 11]
```diff
@@ -510,7 +587,17 @@ def _get_separator(self, name):
         return self.SEPARATORS[name]
 
     def _get_srid(self):
-        return int(self.settings["coordinate_reference_system"].split("_")[1])
+        coordinate_reference_system = self.settings["coordinate_reference_system"]
+        try:
+            return int(coordinate_reference_system.split("_")[1])
+        except (IndexError, ValueError) as err:
+            msg = f"SRID malformed: {coordinate_reference_system}"
+            logger.info(msg)
```
Comment (jrmi) [inline django_geosource/models.py:594]: etc...

File: django_geosource/models.py [hunk 12]
```diff
@@ -510,7 +587,17 @@ def _get_separator(self, name):
         return self.SEPARATORS[name]
 
     def _get_srid(self):
-        return int(self.settings["coordinate_reference_system"].split("_")[1])
+        coordinate_reference_system = self.settings["coordinate_reference_system"]
+        try:
+            return int(coordinate_reference_system.split("_")[1])
+        except (IndexError, ValueError) as err:
+            msg = f"SRID malformed: {coordinate_reference_system}"
```
Comment (jrmi) [inline django_geosource/models.py:593]:
```suggestion
            msg = f"Invalid SRID: {coordinate_reference_system}"
```

File: django_geosource/serializers.py [hunk 13]
```diff
@@ -267,6 +299,25 @@ class Meta:
...
+        url = (
+            data.get("url", "")
+            .replace("{z}", "1")
+            .replace("{y}", "1")
+            .replace("{x}", "1")
+        )
+        try:
+            r = requests.get(url)
+        except requests.ConnectionError:
+            raise ValidationError("The url is not valid")
```
Comment (jrmi) [inline django_geosource/serializers.py:315]:
```suggestion
            raise ValidationError("Can't reach specified tile server. Check your url.")
```

File: django_geosource/serializers.py [hunk 14]
```diff
@@ -267,6 +299,25 @@ class Meta:
...
+
+        if not r.status_code == 200:
+            raise ValidationError("The url is not valid.")
```
Comment (jrmi) [inline django_geosource/serializers.py:318]:
```suggestion
            raise ValidationError("The tile server response is invalid (status code = {r.status_code}).")
```

File: django_geosource/serializers.py [hunk 15]
```diff
@@ -350,7 +401,31 @@ def to_representation(self, obj):
...
+        except (ValueError, GDALException) as err:
+            raise ValidationError(err.args[0])
+
+        # validating id in records
+        for row in records:
+            try:
+                identifier = row[instance.id_field]
+            except KeyError:
+                raise ValidationError(
+                    "Can't find identifier field in one or more records"
```
Comment (jrmi) [inline django_geosource/serializers.py:424]: Can you add record index here ?

**Review by jrmi — CHANGES_REQUESTED** [review]

File: django_geosource/models.py (hunk 1 above)
Comment (CFarcy) [reply django_geosource/models.py]: Done.

File: django_geosource/models.py (hunk 2 above)
Comment (CFarcy) [reply django_geosource/models.py:132]: done

File: django_geosource/models.py (hunk 3 above)
Comment (CFarcy) [reply django_geosource/models.py]: Done

File: django_geosource/models.py (hunk 4 above)
Comment (CFarcy) [reply django_geosource/models.py]: I removed logger

File: django_geosource/models.py (hunk 11 above)
Comment (CFarcy) [reply django_geosource/models.py]: Logger was removed

File: django_geosource/serializers.py (hunk 15 above)
Comment (CFarcy) [reply django_geosource/serializers.py:426]: Done

**Comment by jrmi** [comment]
Can you rebase and squash style or message commit PLS ?

File: django_geosource/models.py (hunk 8 above)
Comment (CFarcy) [reply django_geosource/models.py]: Logger has been removed.

File: django_geosource/models.py (hunk 10 above)
Comment (CFarcy) [reply django_geosource/models.py]: Logger has been removed.

File: django_geosource/models.py (hunk 7 above)
Comment (CFarcy) [reply django_geosource/models.py]: Done.

**Review by jrmi — APPROVED** [review]

**Pull request merged**